### PR TITLE
DAOS-9654 obj: a few refines of iov_buf_len/iov_len

### PR DESCRIPTION
**Pull request by liuxuezhao**
User may provide fetch buffer iov with zero or less iov_len, so need to check
iov_buf_len rather than iov_len in daos_iov_left().
And user possibly provide update buffer sgl/iov with iov_len < iov_buf_len,
to avoid bulk_create's confuse, in that case dup a new sgl and set iov_buf_len
as iov_len.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Comment by liuxuezhao** [comment]
The CI failure of "daos_test -t" due to DAOS-9658.

File: src/include/daos/common.h [hunk 1]
```diff
@@ -339,7 +339,7 @@ daos_sgl_buf_extend(d_sg_list_t *sgl, int idx, size_t new_size);
 	} while (0)
 /** Get the leftover space in an iov of sgl */
 #define daos_iov_left(sgl, iov_idx, iov_off)				\
-	((sgl)->sg_iovs[iov_idx].iov_len - (iov_off))
```
Comment (NiuYawei) [inline src/include/daos/common.h:342]: Is this daos_iov_left() to calculate left data length or left buffer length? It looks to me some places used this function to calculate left data length, and 'iov_len' should be used if that's true?

Comment (liuxuezhao) [reply src/include/daos/common.h:342]: I agree it is kind of confuse
now for update we always set "iov_len == iov_buf_len == buffer_length", 
for fetch user is allowed to only set "iov_buf_len == buffer_length" and possibly set iov_len = 0.
So use iov_buf_len here seems fine? or you have better idea? need to differentiate fetch or update? (that seems not very convenient as well).

Comment (liuxuezhao) [reply src/include/daos/common.h:342]: I added some extra code for obj_update, that if iov_len != iov_buf_len then duplicate the sgl, as what we discussed, thx.

**Review by NiuYawei — APPROVED** [review]

**Comment by liuxuezhao** [comment]
ping reviewer (Niu, what changed after your last review is in commit 3), thanks.

**Review by NiuYawei — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Pull request merged**